SourceCode/Homework05/CRTRenderer.cpp

```diff
@@ -0,0 +1,44 @@
+#include "CRTRenderer.h"
+#include <algorithm>
+#include <iostream>
+
+
+CRTRenderer::CRTRenderer(const CRTScreen& screen, const CRTVector& cameraPosition) : screen(screen), cameraPosition(cameraPosition) 
+{}
+
+CRTImage CRTRenderer::renderScene(const CRTScene& scene, 
+    std::function<CRTColor(const CRTVector&, const CRTTriangle&, float)> visualization) const 
+{
+    const unsigned imageHeight = screen.getHeight();
+    const unsigned imageWidth = screen.getWidth();
+
+    CRTImage image(imageHeight, std::vector<CRTColor>(imageWidth));
+
+    CRTRayMatrix rays = screen.generateRays(cameraPosition);
+
+    std::vector<std::pair<float, const CRTTriangle*>> triangleDistancePair;
+    triangleDistancePair.reserve(scene.size());
+    for (const CRTTriangle& t : scene) {
+        float distance = t.distanceToPoint(cameraPosition);
+        triangleDistancePair.emplace_back(distance, &t);
+    }
+    std::sort(triangleDistancePair.begin(), triangleDistancePair.end(), 

```

Comment:
You could go without a sort, which could be very slow. You just need to track the smallest t when you loop the triangles for the intersection